### ISSUE TITLE
Visibility quota management not working

### ISSUE THREAD

**Issue by binchensun** [lwa-project/ovro_data_recorder#38]
The visibility quota management is struggling. Indeed, many nodes now have disk usage way above their quota. For example, slow visibilities under /data07/slow/ now occupy 9.8 TB of space, which well exceeds the slow disk quota (~6 TB). For fast visibilities, /fast/pipeline/fast on lwacalim01 has 2.4 TB, much much greater than the supposed quota (~500 GB). Observations cannot continue without a working disk quota management. @yupinghuang pointed out that there have been a handful of code changes since a few weeks ago, which may have impacted this.

@jaycedowell, could you please help look into this? Also allow me to @caseyjlaw

**Comment by yupinghuang** [comment]
wha...

```
(base) [yuping@lwacalim08 ~]$ ps aux | grep dr-vfast
pipeline    4537  0.0  0.0 5831928 139456 ?      Ssl  Oct11   0:15 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.82 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-14.lwacalim08.log
pipeline    4539  0.0  0.0 5831888 138912 ?      Ssl  Oct11   0:13 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.78 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-6.lwacalim08.log
pipeline    4542  0.0  0.0 5831876 142632 ?      Ssl  Oct11   0:13 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.81 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-11.lwacalim08.log
pipeline    4543  0.0  0.0 5831948 139924 ?      Ssl  Oct11   0:12 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.79 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-8.lwacalim08.log
pipeline    4544  0.0  0.0 5831880 140184 ?      Ssl  Oct11   0:13 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.77 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-3.lwacalim08.log
pipeline    4545  0.0  0.0 5831968 140376 ?      Ssl  Oct11   0:13 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.76 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-2.lwacalim08.log
pipeline    4558  0.0  0.0 5831864 143332 ?      Ssl  Oct11   0:14 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.80 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-9.lwacalim08.log
pipeline    4573  0.0  0.0 5831408 142868 ?      Ssl  Oct11   0:15 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.81 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-12.lwacalim08.log
pipeline    4596  0.0  0.0 5831972 143024 ?      Ssl  Oct11   0:15 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.80 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-10.lwacalim08.log
pipeline    4612  0.0  0.0 5831924 137372 ?      Ssl  Oct11   0:12 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.79 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-7.lwacalim08.log
pipeline    4624  0.0  0.0 5831884 146492 ?      Ssl  Oct11   0:18 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.78 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-5.lwacalim08.log
pipeline    4639  0.0  0.0 5831864 142508 ?      Ssl  Oct11   0:14 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.77 --port 11002 --cores 68,69,70,71,72,73 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 0 --no-tar --logfile /fast/pipeline/log/dr-vfast-4.lwacalim08.log
pipeline    4643  0.0  0.0 5831864 140652 ?      Ssl  Oct11   0:14 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.76 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-1.lwacalim08.log
pipeline    4650  0.0  0.0 5831924 139412 ?      Ssl  Oct11   0:14 python3 /home/pipeline/proj/lwa-shell/ovro_data_recorder/dr_visibilities.py --address 10.41.0.82 --port 11001 --cores 62,63,64,65,66,67 --quick --nint-per-file 100 --record-directory /fast/pipeline/fast --record-directory-quota 500GB --no-tar --logfile /fast/pipeline/log/dr-vfast-13.lwacalim08.log
```

**Comment by jaycedowell** [comment]
Yes, that has also started happening but only with the vis. recorder services.  I found one of the calim nodes this morning listing two instances of a particular vfast service but systemd reported that the service had failed.

**Comment by yupinghuang** [comment]
This is a fun one. I think our friend here https://github.com/lwa-project/ovro_data_recorder/commit/caafe0720c0c7ebe6c3dd6872130f7d305069ed4 is causing troubles. Removing it does seem to bring quota management back.

Took out the line for vfast5 just now. See below. Also without the grpc fork env thing only one vfast process is around, as opposed to two. So I something probably went wrong when `subprocess.Popen` did a `fork` (or was it `vfork`???) and some deadlocking happened because of grpc's `fork()` handler....

Exhibit A:
Looks like starting on 10/11 StorageLogger stopped being "Done" when SIGTERM is sent. 

```
(base) [pipeline@lwacalim03 log]$ grep -i 'done' dr-vfast-5.lwacalim03.log.2023-09*
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] StorageLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] GlobalLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-16:2023-09-16 04:44:55 [INFO    ] All done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:18:59 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:18:59 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:18:59 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:18:59 [INFO    ] StorageLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:19:00 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:19:00 [INFO    ] GlobalLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-22:2023-09-22 01:19:00 [INFO    ] All done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:46 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:46 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:46 [INFO    ] StorageLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:46 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:47 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:47 [INFO    ] GlobalLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-26 02:15:47 [INFO    ] All done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:55 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:55 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:55 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:56 [INFO    ] StorageLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:56 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:56 [INFO    ] GlobalLogger - Done
dr-vfast-5.lwacalim03.log.2023-09-27:2023-09-27 01:12:56 [INFO    ] All done
(base) [pipeline@lwacalim03 log]$ grep -i 'done' dr-vfast-5.lwacalim03.log.2023-10*
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:31 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:31 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:31 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:31 [INFO    ] StorageLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:31 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:32 [INFO    ] GlobalLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-05:2023-10-05 18:05:32 [INFO    ] All done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 14:44:41 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 14:44:41 [INFO    ] PerformanceLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 14:44:42 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 14:44:42 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 17:25:23 [INFO    ] StatusLogger - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 17:25:23 [INFO    ] StatisticsOp - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 17:25:23 [INFO    ] WriterOp - Done
dr-vfast-5.lwacalim03.log.2023-10-12:2023-10-11 17:25:23 [INFO    ] PerformanceLogger - Done
```

**Comment by jaycedowell** [comment]
Interesting.  I never would have thought that setting that environment variable would cause things to behave so badly.

**Comment by jaycedowell** [comment]
Fixed with 4cc6401.
